src/compiler/checker.ts

```diff
@@ -2595,7 +2595,14 @@ namespace ts {
 
             // Use the type of the initializer expression if one is present
             if (declaration.initializer) {
-                return checkExpressionCached(declaration.initializer);
+                let mapper: TypeMapper;
+                if (declaration.kind === SyntaxKind.PropertyDeclaration) {
+                    const type = getTypeOfBasePropertyDeclaration(<PropertyDeclaration>declaration);
+                    if (type) {
+                        mapper = createTypeMapper([undefinedType, nullType], [type, type]);

```

Comment:
I'm not so sure this works -  doesn't this map *all* undefined and null types in the expression? So if I have

```ts
class Base {
    a: (x: string) => void;
}

class Derived extends Base {
    a = (x = null) => {}
}
```

won't `x` in `Derived` get the type `(x: string) => void` meaning `a` will have the type `(x: (x: string) => void) => void`.

Maybe I misunderstood.